lib/eventstream_rpc.ts

```diff
@@ -0,0 +1,1153 @@
+ */
+
+/**
+ * @packageDocumentation
+ * @module eventstream_rpc
+ */
+
+
+import {CrtError, eventstream, io, cancel} from 'aws-crt';
+import {EventEmitter} from 'events';
+import {toUtf8} from "@aws-sdk/util-utf8-browser";
+
+/**
+ * @internal
+ *
+ * Service model data about an individual operation
+ */
+export interface EventstreamRpcServiceModelOperation {
+    requestShape: string,
+    responseShape: string,
+    outboundMessageShape?: string,
+    inboundMessageShape?: string,
+    errorShapes: Set<string>
+}
+
+/**
+ * @internal
+ *
+ * Normalizers strip unmodeled data from a value (by making a new value with only modeled data)
+ */
+export type ShapeNormalizer = (value: any) => any;
+
+/**
+ * @internal
+ *
+ * Validators throw errors if modeled data is of the wrong type, is missing when required, or violates any
+ * other checkable protocol constraint we can think of.
+ *
+ * Validation is only performed on input shapes.  Output shapes must be left alone in order to support
+ * service evolution indepedent of client.
+ */
+export type ShapeValidator = (value: any) => void;
+
+/**
+ * @internal
+ *
+ * Deserializers take a raw eventstream message and return a modeled shape.  If a modeled shape could not be
+ * produced, an error is thrown.
+ */
+export type ShapeDeserializer = (message: eventstream.Message) => any;
+
+/**
+ * @internal
+ *
+ * Serializers take a modeled shape and return an eventstream message.  The assumption is that once a value
+ * has been both validated and normalized, serializing the final value will result in a message that
+ * an eventstream_rpc server can deserialize without error.
+ */
+export type ShapeSerializer = (value: any) => eventstream.Message;
+
+/**
+ * @internal
+ *
+ * Collection of service-specific utility functions and definitions that enable an eventstream RPC client to
+ * correctly perform operations
+ */
+export interface EventstreamRpcServiceModel {
+
+    normalizers: Map<string, ShapeNormalizer>;
+
+    validators: Map<string, ShapeValidator>;
+
+    deserializers: Map<string, ShapeDeserializer>;
+
+    serializers: Map<string, ShapeSerializer>;
+
+    operations: Map<string, EventstreamRpcServiceModelOperation>;
+
+    enums: Map<string, Set<string>>;
+}
+
+/**
+ * Indicates the general category of an error thrown by the eventstream RPC implementation
+ */
+export enum RpcErrorType {
+
+    /**
+     * An error occurred while serializing a client model into a message in the eventstream protocol.
+     */
+    SerializationError,
+
+    /**
+     * An error occurred while deserializing a message from the eventstream protocol into the client model.
+     */
+    DeserializationError,
+
+    /**
+     * An error occurred during the connect-connack handshake between client and server.  Usually this means
+     * the connect was not accepted by the server and thus hints at an authentication problem.
+     */
+    HandshakeError,
+
+    /**
+     * An error that isn't classifiable occurred.
```

Comment:
Nitpick: change to `... isn't classifiable by one of the other RpcErrorTypes occurred`?